### PR TITLE
[AMDGPU] Use `bf16` instead of `i16` for bfloat

### LINKED ISSUE #79369: [AMDGPU] Incorrect parsing of bf16 literals
bf16 immediate operands are not handled correctly by asm parser (but seem OK in the codegen):
```
llvm-mc -arch=amdgcn -mcpu=gfx1200 -show-encoding <<< 'v_dot2_bf16_bf16 v5, v1, v2, 100.0'
v_dot2_bf16_bf16 v5, v1, v2, 0x5640     ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0xfe,0x03,0x40,0x56,0x00,0x00]
```
bf16 constants are essentially fp32 with all zero low 16 bits. So 100.0 shall be encoded as 0x42c80000, and since we only accept 16 bits in the asm hex for it has to be 0x42c8.
```
llvm-mc -arch=amdgcn -mcpu=gfx1200 -show-encoding <<< 'v_dot2_bf16_bf16 v5, v1, v2, 1.0'
v_dot2_bf16_bf16 v5, v1, v2, 0x3c00     ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0xfe,0x03,0x00,0x3c,0x00,0x00]
```
This shall be inline immediate.

Basically we are parsing bf16 constants as f16.

### PR DESCRIPTION
**Pull request by shiltian**
Currently we generally use `i16` to represent `bf16` in those tablegen files. This patch is trying to use `bf16` directly.

Fix #79369.

### CODE REVIEW

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 1]
```diff
@@ -2819,11 +2819,11 @@ def int_amdgcn_fdot2_f16_f16 :
 def int_amdgcn_fdot2_bf16_bf16 :
   ClangBuiltin<"__builtin_amdgcn_fdot2_bf16_bf16">,
   DefaultAttrsIntrinsic<
-    [llvm_i16_ty],   // %r
+    [llvm_bfloat_ty],   // %r
```
Comment (rampitec) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:2822]: This is the change in the existing interface, it will break programs already using it. It also will break clang builtin.

You can get away with changing only instruction's profile to use BF16 and tweak pattern with casts to bf16.

Comment (arsenm) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td]: It will just maybe require changing the clang builtin codegen to insert the casts. You still need to make this change

Comment (arsenm) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td]: You should also add IR auto upgrade

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp [hunk 2]
```diff
@@ -1562,8 +1562,9 @@ bool IRTranslator::translateBitCast(const User &U,
 
 bool IRTranslator::translateCast(unsigned Opcode, const User &U,
                                  MachineIRBuilder &MIRBuilder) {
-  if (U.getType()->getScalarType()->isBFloatTy() ||
-      U.getOperand(0)->getType()->getScalarType()->isBFloatTy())
+  if (Opcode != TargetOpcode::G_BITCAST &&
```
Comment (shiltian) [inline llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:1565]: I think bitcast can still be supported even if we don't know how to convert from bfloat to other types mentioned in #71470. If so, we might want a separate patch for this.

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td (hunk 1 above)
Comment (shiltian) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td:2822]: Does it make sense if we also update both the Clang builtins and LLVM builtins here, and use IR auto upgrade to rewrite existing uses?

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:1565]: Unrelated, but this also misses hidden casts implied by the ABI

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td (hunk 1 above)
Comment (arsenm) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td:2822]: Changing the clang builtin breaks user code. Anyone using it will now have an implicit itofp cast to bfloat. For the first step, I would leave the clang builtin alone (and only make the codegen change to emit the new IR). We would either need to introduce a new builtin with the corrected signature, or add a language flag to change the behavior

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:1565]: This change should be done separately

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 3]
```diff
@@ -2835,8 +2835,8 @@ def int_amdgcn_fdot2_f32_bf16 :
   DefaultAttrsIntrinsic<
     [llvm_float_ty], // %r
     [
-      llvm_v2i16_ty, // %a
-      llvm_v2i16_ty, // %b
+      llvm_v2bf16_ty, // %a
+      llvm_v2bf16_ty, // %b
```
Comment (arsenm) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:2839]: For potential revert pain minimization, I would do each intrinsic in a separate commit

File: llvm/test/MC/AMDGPU/bf16_imm.s [hunk 4]
```diff
@@ -0,0 +1,8 @@
+// RUN: llvm-mc -arch=amdgcn -mcpu=gfx1100 -show-encoding %s | FileCheck %s
+// RUN: llvm-mc -arch=amdgcn -mcpu=gfx1200 -show-encoding %s | FileCheck %s
+
+v_dot2_bf16_bf16 v5, v1, v2, 100.0
```
Comment (arsenm) [inline llvm/test/MC/AMDGPU/bf16_imm.s:4]: does this help with #79369 at all?

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 5]
```diff
@@ -5908,8 +5908,6 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
           }
         }
 
-        assert(ArgValue->getType()->canLosslesslyBitCastTo(PTy) &&
```
Comment (shiltian) [inline clang/lib/CodeGen/CGBuiltin.cpp:5911]: This change might need to go to a separate patch.

File: llvm/test/MC/AMDGPU/bf16_imm.s [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+// CHECK: v_dot2_bf16_bf16 v5, v1, v2, 0x42c8 ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0xfe,0x03,0xc8,0x42,0x00,0x00]
+
+v_dot2_bf16_bf16 v5, v1, v2, 1.0
+// v_dot2_bf16_bf16 v5, v1, v2, 0x3f80 ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0xfe,0x03,0x80,0x3f,0x00,0x00]
```
Comment (rampitec) [inline llvm/test/MC/AMDGPU/bf16_imm.s:8]: FYI: this shall be inline literal. I.e:
    0xd6672005
    0x03ca0501

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp (hunk 2 above)
Comment (rampitec) [reply llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:1565]: This is actually an orthogonal problem. Global ISel is completely broken for bf16 and whatever the outcome of the supporting bf16 in codegen is we just need to be ready some gisel tests will fail and will need to be disabled.

File: llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUInstPrinter.cpp [hunk 7]
```diff
@@ -521,8 +521,11 @@ void AMDGPUInstPrinter::printImmediateV216(uint32_t Imm, uint8_t OpType,
     if (printImmediateFloat32(Imm, STI, O))
       return;
     break;
+  case AMDGPU::OPERAND_REG_IMM_V2BF16:
   case AMDGPU::OPERAND_REG_IMM_V2FP16:
+  case AMDGPU::OPERAND_REG_INLINE_C_V2BF16:
   case AMDGPU::OPERAND_REG_INLINE_C_V2FP16:
+  case AMDGPU::OPERAND_REG_INLINE_AC_V2BF16:
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUInstPrinter.cpp:528]: It does not seem right, and there are no tests for v2bf16 added. I am not sure though we have instructions which can accept this type of operand.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 8]
```diff
@@ -4181,13 +4181,20 @@ bool SIInstrInfo::isInlineConstant(const MachineOperand &MO,
   case AMDGPU::OPERAND_REG_INLINE_C_V2INT16:
   case AMDGPU::OPERAND_REG_INLINE_AC_V2INT16:
     return AMDGPU::isInlinableLiteralV2I16(Imm);
+  case AMDGPU::OPERAND_REG_IMM_V2BF16:
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4184]: It does not seem isInlinableLiteralV2F16() can handle bf16.

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll [hunk 9]
```diff
@@ -79,17 +79,17 @@ define amdgpu_ps void @test_llvm_amdgcn_fdot2_bf16_bf16_sis(
 ; GFX11:       ; %bb.0: ; %entry
 ; GFX11-NEXT:    v_mov_b32_e32 v2, s1
 ; GFX11-NEXT:    s_delay_alu instid0(VALU_DEP_1)
-; GFX11-NEXT:    v_dot2_bf16_bf16 v2, s0, 0x10001, v2
+; GFX11-NEXT:    v_dot2_bf16_bf16 v2, s0, 0x3f803f80, v2
```
Comment (rampitec) [inline llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll:82]: This shall be encoded as inline immediate 1.0.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 8 above)
Comment (shiltian) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4184]: Yeah, I made some mistakes here. Will take care of them.

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll (hunk 9 above)
Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll:82]: Yeah, but I tested the FP16 version `llvm.amdgcn.fdot2.f16.f16` (w/ trunk w/o my patch), it generates `v_dot2_f16_f16 v2, s0, 0x3c003c00, v2`. I think we generally have issues with showing inline literals.

File: llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp [hunk 10]
```diff
@@ -2730,6 +2749,12 @@ std::optional<unsigned> getInlineEncodingV2I16(uint32_t Literal) {
   return getInlineEncodingV216(false, Literal);
 }
 
+// Encoding of the literal as an inline constant for a V_PK_*_BF16 instruction
+// or nullopt.
+std::optional<unsigned> getInlineEncodingV2BF16(uint32_t Literal) {
+  return getInlineEncodingV216(true, Literal);
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:2755]: This part is still WIP along with https://github.com/llvm/llvm-project/pull/81345.

File: llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp [hunk 11]
```diff
@@ -2660,15 +2660,34 @@ bool isInlinableLiteral16(int16_t Literal, bool HasInv2Pi) {
...
-         Val == 0x3800 || // 0.5
-         Val == 0xB800 || // -0.5
-         Val == 0x4000 || // 2.0
-         Val == 0xC000 || // -2.0
-         Val == 0x4400 || // 4.0
-         Val == 0xC400 || // -4.0
-         Val == 0x3118;   // 1/2pi
+
+  // FP16
+  if (Val == 0x3C00 || // 1.0
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:2665]: This function might be removed eventually in https://github.com/llvm/llvm-project/pull/81345.

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll (hunk 9 above)
Comment (rampitec) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll:82]: Well, this is unrelated to the patch itself. We can use inline 1.0 here, but then we must use op_sel_hi to produce it in the high half.

**Comment by shiltian** [comment]
The patch is in a good shape now. I have made two other prime patches (#81674 and #81669). I'll rebase this one once they are landed.

This patch only changes one bf16 instruction with the necessary infrastructure for others. I'll update all of them once this patch is landed.

However, I don't think `isInlinableLiteral16` works correctly because the encoding of the floating point inline literals are different for fp16 and bf16, but apparently for now it can only recognize fp16 encoding. This patch at least makes the asm printer work properly. #81345 is trying to fix it correctly, but that is unrelated to this patch.

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll (hunk 9 above)
Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll:82]: FWIW, #81345 can solve the issue but I'm struggling with getting two test cases passed.

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll [hunk 12]
```diff
@@ -1,8 +1,7 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
 ; RUN: llc -mtriple=amdgcn -mcpu=gfx1100 -verify-machineinstrs < %s | FileCheck %s --check-prefixes=GFX11,SDAG-GFX11
-; RUN: llc -global-isel -mtriple=amdgcn -mcpu=gfx1100 -verify-machineinstrs < %s | FileCheck %s --check-prefixes=GFX11,GISEL-GFX11
```
Comment (shiltian) [inline llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll:3]: This change is because of the discussion here (https://github.com/llvm/llvm-project/pull/80908/files#r1483394728).

File: llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUInstPrinter.cpp [hunk 13]
```diff
@@ -488,6 +488,49 @@ static bool printImmediateFloat16(uint32_t Imm, const MCSubtargetInfo &STI,
...
+  if (isInlinableIntLiteral(SImm)) {
+    O << SImm;
+    return;
+  }
+
+  uint16_t HImm = static_cast<uint16_t>(Imm);
+  if (printImmediateBFloat16(HImm, STI, O))
+    return;
+
+  uint64_t Imm16 = static_cast<uint16_t>(Imm);
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUInstPrinter.cpp:530]: It's the same as HImm above.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 14]
```diff
@@ -4185,9 +4185,17 @@ bool SIInstrInfo::isInlineConstant(const MachineOperand &MO,
...
+  case AMDGPU::OPERAND_REG_INLINE_C_V2BF16:
+  case AMDGPU::OPERAND_REG_INLINE_AC_V2BF16:
+    return AMDGPU::isInlinableLiteralV2BF16(Imm);
+  case AMDGPU::OPERAND_REG_IMM_BF16:
   case AMDGPU::OPERAND_REG_IMM_FP16:
+  case AMDGPU::OPERAND_REG_IMM_BF16_DEFERRED:
   case AMDGPU::OPERAND_REG_IMM_FP16_DEFERRED:
+  case AMDGPU::OPERAND_REG_INLINE_C_BF16:
   case AMDGPU::OPERAND_REG_INLINE_C_FP16:
+  case AMDGPU::OPERAND_REG_INLINE_AC_BF16:
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4198]: It seems isInlinableLiteral16() cannot handle bf16?

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll (hunk 12 above)
Comment (rampitec) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll:3]: Change 'RUN' with 'XUN' and add a comment instead.

File: llvm/test/MC/AMDGPU/bf16_imm.s [hunk 15]
```diff
@@ -0,0 +1,8 @@
+// RUN: llvm-mc -arch=amdgcn -mcpu=gfx1100 -show-encoding %s | FileCheck %s
```
Comment (rampitec) [inline llvm/test/MC/AMDGPU/bf16_imm.s:1]: You also need a disasm test for this.

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td (hunk 1 above)
Comment (rampitec) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td:2822]: clang/test/CodeGenOpenCL/builtins-amdgcn-dl-insts-gfx11.cl fails. You need to insert casts to bf16 while lowering it to make it working.

Comment (shiltian) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td:2822]: The cast will be inserted automatically in `clang/lib/CodeGen/CGBuiltin.cpp` after removing the two assertions. I reverted my change to the test case by accident. Lol

File: llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUInstPrinter.cpp (hunk 13 above)
Comment (shiltian) [reply llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUInstPrinter.cpp:530]: Yeah, but it is to make the type promoted to `uint64_t` w/o any ambiguity.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 14 above)
Comment (shiltian) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4198]: Yes, `isInlinableLiteral16` can't because it can't tell `fp16` and `bf16` apart by just looking at the value. That's the reason I tried really hard to get rid of `isInlinableLiteral16` in #81345 and favors the explicit version.

Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4198]: But right in this place you know the actual format. So you can split F16 and BF16 code and call different functions.

Comment (shiltian) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp]: Okay, I copied `isInlinableLiteralBF16` from https://github.com/llvm/llvm-project/pull/81345. Initially I thought https://github.com/llvm/llvm-project/pull/81345 would be landed before this one, but I went down that rabbit hole and still didn't make it fully work.

File: llvm/test/MC/Disassembler/AMDGPU/bf16_imm.txt [hunk 16]
```diff
@@ -0,0 +1,8 @@
+# RUN: llvm-mc -triple=amdgcn -mcpu=gfx1100 -disassemble -show-encoding < %s | FileCheck %s
+# RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -disassemble -show-encoding < %s | FileCheck %s
+
+# CHECK: v_dot2_bf16_bf16 v5, v1, v2, 0x42c8
+0x05,0x00,0x67,0xd6,0x01,0x05,0xfe,0x03,0xc8,0x42,0x00,0x00
+
+# CHECK: v_dot2_bf16_bf16 v5, v1, v2, 0x3c00
```
Comment (shiltian) [inline llvm/test/MC/Disassembler/AMDGPU/bf16_imm.txt:7]: The disassembler doesn't work properly because when it sees `242` with bitwidth 16, it doesn't know whether it is bf16 or fp16.

File: llvm/test/MC/Disassembler/AMDGPU/bf16_imm.txt [hunk 17]
```diff
@@ -0,0 +1,8 @@
+# RUN: llvm-mc -triple=amdgcn -mcpu=gfx1100 -disassemble -show-encoding < %s | FileCheck %s
+# RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -disassemble -show-encoding < %s | FileCheck %s
+
+# CHECK: v_dot2_bf16_bf16 v5, v1, v2, 0x42c8
```
Comment (rampitec) [inline llvm/test/MC/Disassembler/AMDGPU/bf16_imm.txt:4]: Add encoding to the check lines. Currently it is broken and encoded value is different from decoded one.

**Comment by shiltian** [comment]
I'll create a ticket about the decoder.

File: llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCCodeEmitter.cpp [hunk 18]
```diff
@@ -157,6 +157,27 @@ static uint32_t getLit16Encoding(uint16_t Val, const MCSubtargetInfo &STI) {
   return 255;
 }
 
+static uint32_t getLitBF16Encoding(uint16_t Val) {
+  uint16_t IntImm = getIntInlineImmEncoding(static_cast<int16_t>(Val));
+  if (IntImm != 0)
+    return IntImm;
+
+  // clang-format off
+  switch (Val) {
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCCodeEmitter.cpp:166]: Can this call `getInlineEncodingV2BF16`?

Comment (shiltian) [reply llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCCodeEmitter.cpp:166]: In theory, yes, but for now we can't because `getInlineEncodingV2BF16` can't handle some cases (that I didn't dig yet). It looks like in the conversion between `uint16_t` and `uint32_t` that makes some test cases fail. IMO we need to unify them (not only for 16-bit) in one place instead of having almost the same logic at least in three places.

Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCCodeEmitter.cpp:166]: Yeah, I really don't like having 4 different copies of this list of hex values (0x3f00, 0xbf00...).

File: llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp [hunk 19]
```diff
@@ -2652,6 +2652,23 @@ bool isInlinableLiteral32(int32_t Literal, bool HasInv2Pi) {
          (Val == 0x3e22f983 && HasInv2Pi);
 }
 
+bool isInlinableLiteralBF16(int16_t Literal, bool HasInv2Pi) {
+  if (!HasInv2Pi)
+    return false;
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:2657]: It does not change the behavior, but generally it shall only matter when you compare value to 0x3E22.

File: llvm/test/MC/AMDGPU/bf16_imm.s [hunk 20]
```diff
@@ -0,0 +1,8 @@
...
+// RUN: llvm-mc -arch=amdgcn -mcpu=gfx1200 -show-encoding %s | FileCheck %s
+
+v_dot2_bf16_bf16 v5, v1, v2, 100.0
+// CHECK: v_dot2_bf16_bf16 v5, v1, v2, 0x42c8 ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0xfe,0x03,0xc8,0x42,0x00,0x00]
+
+v_dot2_bf16_bf16 v5, v1, v2, 1.0
+// CHECK: v_dot2_bf16_bf16 v5, v1, v2, 1.0 ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0xca,0x03]
```
Comment (rampitec) [inline llvm/test/MC/AMDGPU/bf16_imm.s:8]: Can you add couple more tests here? The same instruction, but with the immediate in place of v2bf16 operand, so we test that code path too. Both immediate and inline literal. Here and in disasm.

Comment (shiltian) [reply llvm/test/MC/AMDGPU/bf16_imm.s:8]: It looks like our assembler doesn't like the case such as `v_dot2_bf16_bf16 v5, v1, 0x42c842c8, 0x42c8`, even w/o this patch. Even the asm (`v_dot2_bf16_bf16 v2, s0, 0x3f803f80, v2`) generated in another file above (`llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll`) can't be recognized by our assembler. 😄

**Review by rampitec — APPROVED** [review]
Thanks. There are definitely at least 2 outstanding problems, but it seems there are no regressions comparing to what we have now. LGTM.
